typify-impl/src/lib.rs

```diff
@@ -219,11 +225,142 @@ impl TypeSpace {
+        // to break with a Box<T>.
+        for index in 0..def_len {
+            let type_id = TypeId(base_id + index);
+
+            let mut box_id = None;
+
+            let mut type_entry = self.id_to_entry.get_mut(&type_id).unwrap().clone();
+            self.break_trivial_cyclic_refs(&type_id, &mut type_entry, &mut box_id);
+            let _ = self.id_to_entry.insert(type_id, type_entry);
+        }
 
         Ok(())
     }
 
+    /// If a type refers to itself, this creates a cycle that will eventually be
+    /// emit as a Rust struct that cannot be constructed. Break those cycles
+    /// here.
+    ///
+    /// While we aren't yet handling the general case of type containment
+    /// cycles, it's not that bad to look at trivial cycles such as:
+    ///
+    ///   1) A type refering to itself: A -> A
+    ///   2) A type optionally referring to itself: A -> Option<A>
+    ///   3) An enum variant referring to itself, either optionally or directly.
+    ///
+    /// TODO currently only trivial cycles are broken. A more generic solution
+    /// may be required, but it may also a point to ask oneself why such a
+    /// complicated type is required :) A generic solution is difficult because
+    /// certain cycles introduce a question of *where* to Box to break the
+    /// cycle, and there's no one answer to this.
+    ///
+    fn break_trivial_cyclic_refs(
+        &mut self,
+        type_id: &TypeId,
+        type_entry: &mut TypeEntry,
+        box_id: &mut Option<TypeId>,
+    ) {
+        match &mut type_entry.details {
+            // Look for the case where an option refers to the parent type
+            TypeEntryDetails::Option(option_type_id) => {
+                if *option_type_id == *type_id {
+                    *option_type_id = box_id
+                        .get_or_insert_with(|| self.id_to_box(type_id))
+                        .clone();
+                }
+            }
+
+            // Look for the case where a struct property refers to the parent
+            // type
+            TypeEntryDetails::Struct(s) => {
+                for prop in &mut s.properties {
+                    if prop.type_id == *type_id {
+                        // A struct property directly refers to the parent type
+                        prop.type_id = box_id
+                            .get_or_insert_with(|| self.id_to_box(type_id))
+                            .clone();
+                    } else {
+                        // A struct property optionally refers to the parent type
+                        let mut prop_type_entry =
+                            self.id_to_entry.get_mut(&prop.type_id).unwrap().clone();
+                        self.break_trivial_cyclic_refs(&type_id, &mut prop_type_entry, box_id);
+                        let _ = self
+                            .id_to_entry
+                            .insert(prop.type_id.clone(), prop_type_entry);
+                    }
```

Comment:
Thanks for the prompt - I originally assumed I could not pull that into a method because it wouldn't make the borrow checker happy, but turns out it's possible. Done in 3d9210c